### PR TITLE
QByteArray support

### PR DESCRIPTION
**Pull request by geigertom**
Added support for `QByteArray`.

Create `QByteArray` in .NET using `byte[]`:
```
[NotifySignal("byteArrayPropertyChanged")]
public byte[] ByteArrayProperty { get; set; } = new byte[]{1,3,6,8};
```

Receive `QByteArray` in .NET:
```
public void Process(dynamic o) {
    byte[] bytes = (byte[])o;
    Console.WriteLine("Received byte array size: " + bytes.Length);
    ...
}
```

In QML, there is no direct way to access `QByteArray` objects, they are, however, implicitly converted to javascript `ArrayBuffer` objects [see: [Qt docs](https://doc.qt.io/qt-5/qtqml-cppintegration-data.html#qbytearray-to-javascript-arraybuffer)] which can be used like this:

```
function logValue() {
    var data = new Uint8Array(netObject.byteArrayProperty)
    console.log("first element from byte array: " + data[0]);
}
```

### CODE REVIEW

**Comment by vadi2** [comment]
Could you update the docs as well?

**Comment by geigertom** [comment]
Hey, do you mean the README or is there any more documention that I'm not aware of?

**Comment by vadi2** [comment]
Yeah there is https://qmlnet.github.io which is managed in https://github.com/qmlnet/qmlnet.github.io. Stuffing everything into the README would probably be too much, I agree!

**Comment by geigertom** [comment]
Ah you're right, I will add some documentation there as soon as the PR is accepted and merged.

**Comment by pauldotknopf** [comment]
I'm not ignoring you. I just haven't had time lately. I will get to it this week.

File: src/net/Qml.Net/Internal/Qml/NetVariant.cs [hunk 1]
```diff
@@ -88,6 +88,22 @@ public string String
...
+        {
+            get {
+                int count;
+                IntPtr arrayHandle = Interop.NetVariant.GetBytes(Handle, out count);
+                byte[] managedArray = new byte[count];
+                Marshal.Copy(arrayHandle, managedArray, 0, count);
+                return managedArray;
+            } 
+            set {
+                GCHandle pinnedArray = GCHandle.Alloc(value, GCHandleType.Pinned);
```
Comment (pauldotknopf) [inline src/net/Qml.Net/Internal/Qml/NetVariant.cs]: I think you need to free this handle.

File: src/native/QmlNet/QmlNet/qml/NetVariant.cpp [hunk 2]
```diff
@@ -663,6 +676,20 @@ Q_DECL_EXPORT QmlNetStringContainer* net_variant_getString(NetVariantContainer*
     return createString(string);
 }
 
+Q_DECL_EXPORT void net_variant_setBytes(NetVariantContainer* container, LPCSTR value, int count) {
+    if(value == nullptr) {
+        container->variant->setBytes(nullptr);
+    } else {
+        container->variant->setBytes(QByteArray::fromRawData(value, count));
```
Comment (pauldotknopf) [inline src/native/QmlNet/QmlNet/qml/NetVariant.cpp]: ```fromRawData``` does not copy the data. Because of this, we need to copy it ourselves.

**Comment by pauldotknopf** [comment]
Everything looks good!

There are missing tests, but I'd had to block this because of it. I'm sure it's fine. I'll add them later.

**Comment by pauldotknopf** [comment]
Quick update, I'm working on the unit tests now. When it's done, I will fully merge it.

Question, how are you using this project? Are you using .NET Core as entry point, using NuGet libs?

**Comment by pauldotknopf** [comment]
@geigertom, don't worry about the documentation for now. There are many things missing in it at the moment.

**Comment by pauldotknopf** [comment]
Thanks for your contribution!

**Comment by pauldotknopf** [comment]
@all-contributors please add @geigertom for code.

**Comment by geigertom** [comment]
Thanks! I'm using a traditional QT Project as an entry point, as described here: `https://github.com/qmlnet/qmlnet/tree/develop/samples/hosting`

**Comment by pauldotknopf** [comment]
Great! That's the pro way of doing things!

I'm interested in what you are building. Ping me on Gitter if you want to share.

**Comment by pauldotknopf** [comment]
Closes #148

**Comment by geigertom** [comment]
Right now I'm still in the process of prototyping. What I try to archive is using Qt's Camera module to capture real time video and send it, via a QVideoFilter to .NET (hence the need for QByteArray support). In .NET I want to make use of [OpenCVSharp](https://github.com/SciSharp/Numpy.NET) and [Numpy](https://github.com/shimat/opencvsharp) to recognize shapes and objects in the video material. Based on some calculations i want to draw objects to QML Canvas elements. I already built a working POC in Python (PyQt, opencv, numpy, etc.) but I would really love to use .NET, as from my experience using Python for bigger projects becomes really difficult to handle, and I like the tooling provided in .NET. 


There's one thing regarding QByteArray conversion, that I'm uncertain about: After I recently found out, that int[] Properties [can be accessed from QML directly](https://github.com/qmlnet/qmlnet/issues/151), I guess the same goes for byte[] properties. I do not really need that particular use case, but doesn't the implicit conversion of byte[] to QByteArray (.NET->QML) shadow that more lightweight approach in case anyone wants to pass simple byte Arrays directly? Should we support that, or is the overhead negligible?

**Pull request merged**